docs: write docs reflecting how to use yargs with webpack

version:  6.6.0

It look like yargs  cannot be used with webpack ( tested with latest webpack version 2.2)

Here is how to reproduce

 webconfig.pack.js

```javascript

var path = require("path");
module.exports = {

    target: "node",
    context: __dirname,
    node: {
        console: true,
        global: true,
        process: true,
        Buffer: true,
        setImmediate: true,
        __dirname: false,
        __filename: false,

    },
    entry: {
        main: path.resolve(__dirname,"bug.js"),
    },
    output: {
        path: path.resolve(__dirname),
        filename: "[name].dist.js"
    },
 };
 
```

bug.js
```javascript 
var yargs = require("yargs");
```

package.json
```json
{
  "name": "yargs_issue_with_web_pack",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "webpack": "./node_modules/.bin/webpack"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "webpack": "^2.2.1"
  },
  "dependencies": {
    "yargs": "^6.6.0"
  }
}
```

commande:

```
$npm run webpack
```


Output is
```
WARNING in ./~/yargs/index.js
11:39-46 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

WARNING in ./~/yargs/yargs.js
248:33-40 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

WARNING in ./~/yargs/yargs.js
466:71-78 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

WARNING in ./~/yargs-parser/index.js
444:21-48 Critical dependency: the request of a dependency is an expression

WARNING in ./~/require-main-filename/index.js
2:25-32 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
```

running the webpacked scripts 
    $ node main.dist.js


produces
```
C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:662                                                                                    
  const argv = yargs(processArgs, cwd, !(function webpackMissingModule() { var e = new Error("Cannot find module \".\""); e.code = 'MODULE_NOT_FOUND';; thro
                                                                                                                                                        ^   
                                                                                                                                                            
Error: Cannot find module "."                                                                                                                               
    at webpackMissingModule (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:662:84)                                                   
    at Argv (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:662:163)                                                                  
    at Object.module.exports (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:657:1)                                                   
    at __webpack_require__ (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:20:30)                                                     
    at Object.<anonymous> (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:11603:13)                                                   
    at __webpack_require__ (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:20:30)                                                     
    at C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:66:18                                                                           
    at Object.<anonymous> (C:\projects\stuff\yargs_issue_with_web_pack\main.dist.js:69:10)                                                      
    at Module._compile (module.js:570:32)                                                                                                                   
    at Object.Module._extensions..js (module.js:579:10)                                                                                                     
                                                                                                                                                            
```

The workaround proposed above doesn't work for me for both 6.6.0 and current 10.0.3 versions. My environment is the following:

Ubuntu 16.04 LTS
Node 6.11.0
Webpack 3.8.1

These warnings are caused by single import line:

```javascript
import yargs from 'yargs/yargs';
```

For yargs 6.6.0 I get:

```
WARNING in ./node_modules/yargs/yargs.js
248:33-40 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/require-main-filename/index.js
2:25-32 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/require-main-filename/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/yargs-parser/index.js
444:21-48 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs-parser/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/yargs/yargs.js
466:71-78 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js
```

For yargs 10.0.3 I get:

```
WARNING in ./node_modules/yargs/lib/apply-extends.js
28:24-55 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs/lib/apply-extends.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/require-main-filename/index.js
2:25-32 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/require-main-filename/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/yargs/yargs.js
349:33-40 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/yargs/lib/apply-extends.js
43:82-105 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs/lib/apply-extends.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/yargs-parser/index.js
461:21-48 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs-parser/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/yargs/yargs.js
505:71-78 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js

WARNING in ./node_modules/cross-spawn/index.js
Module not found: Error: Can't resolve 'spawn-sync' in '/home/ezze/Development/planner/node_modules/cross-spawn'
 @ ./node_modules/cross-spawn/index.js 32:26-47
 @ ./node_modules/execa/index.js
 @ ./node_modules/yargs/node_modules/os-locale/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./src/main/main.js
```

Same goes for:

```js
import { yargs } from 'yargs';
```

Yargs 8.0.2
Mac Os High Sierra
Node 8.9.1
Webpack 3.8.1

```
WARNING in ./node_modules/yargs/index.js
11:39-46 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/yargs/lib/apply-extends.js
27:24-55 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs/lib/apply-extends.js
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/require-main-filename/index.js
2:25-32 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/require-main-filename/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/yargs/yargs.js
341:33-40 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/yargs/lib/apply-extends.js
42:82-105 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs/lib/apply-extends.js
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/yargs-parser/index.js
449:21-48 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/yargs-parser/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/yargs/yargs.js
497:71-78 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts

WARNING in ./node_modules/cross-spawn/index.js
Module not found: Error: Can't resolve 'spawn-sync' in '/Users/martien/MEGA/Projects/alfred-workflow-todoist/node_modules/cross-spawn'
 @ ./node_modules/cross-spawn/index.js 32:26-47
 @ ./node_modules/execa/index.js
 @ ./node_modules/os-locale/index.js
 @ ./node_modules/yargs/yargs.js
 @ ./node_modules/yargs/index.js
 @ ./index.ts
```

--- a/README.md
+++ b/README.md
@@ -98,6 +98,10 @@ npm i @types/yargs --save-dev
 
 See usage examples in [docs](/docs/typescript.md).
 
+## Webpack
+
+See usage examples of yargs with webpack in [docs](/docs/webpack.md).
+
 ## Community :
 
 Having problems? want to contribute? join our [community slack](http://devtoolscommunity.herokuapp.com).

--- a/docs/webpack.md
+++ b/docs/webpack.md
@@ -1,11 +1,21 @@
 # Webpack usage examples
 
-## Setup
+## Install yargs
 
-### Javascript
+Install yargs as a production dependency, as it will _not_ be included in the pack (not yet compatible):
+```bash
+$ npm install --save yargs
+```
+
+## Javascript setup (see below for typescript)
+
+Install dev dependencies:
+```bash
+$ npm install --save-dev webpack webpack-cli
+```
 
 Create `src/index.js`:
-```ts
+```js
 const yargs = require('yargs')
 
 console.log(yargs.parse())
@@ -14,12 +24,13 @@ console.log(yargs.parse())
 Create `webpack.config.js`:
 ```js
 const path = require('path');
-const nodeExternals = require('webpack-node-externals');
 
 module.exports = {
   entry: './src/index.js',
-  externals: [nodeExternals()],
-  mode: 'production',
+  // Keep yargs as an external dependency
+  externals: {
+    'yargs': 'commonjs yargs'
+  },
   output: {
     path: path.resolve(__dirname, 'dist'),
     filename: 'index.js'
@@ -28,12 +39,12 @@ module.exports = {
 }
 ```
 
-Install dependencies (all as dev-dependencies, as prod will not need any dependency):
-```bash
-$ npm install --save-dev webpack webpack-cli webpack-node-externals yargs
-```
+## Typescript setup
 
-### Typescript
+Install dev dependencies:
+```bash
+$ npm install --save-dev ts-loader typescript webpack webpack-cli @types/yargs
+```
 
 Create `src/index.ts`:
 ```ts
@@ -54,12 +65,13 @@ Create `tsconfig.json`:
 Create `webpack.config.js`:
 ```js
 const path = require('path');
-const nodeExternals = require('webpack-node-externals');
 
 module.exports = {
   entry: './src/index.ts',
-  externals: [nodeExternals()],
-  mode: 'production',
+  // Keep yargs as an external dependency
+  externals: {
+    'yargs': 'commonjs yargs'
+  },
   module: {
     rules: [
       {
@@ -81,20 +93,15 @@ module.exports = {
 }
 ```
 
-Install dependencies (all as dev-dependencies, as prod will not need any dependency):
-```bash
-$ npm install --save-dev ts-loader typescript webpack webpack-cli webpack-node-externals @types/yargs yargs
-```
-
 ## Build
 
 ```bash
-$ ./node_modules/.bin/webpack
+$ ./node_modules/.bin/webpack --mode=production
 ```
 
 ## Run
 
 ```bash
 $ node dist/index.js
-{ _: [], '$0': 'dist\\index.js' }
+{ _: [], '$0': 'dist/index.js' }
 ```

--- a/docs/webpack.md
+++ b/docs/webpack.md
@@ -1,18 +1,18 @@
 # Webpack usage examples
 
-## Install yargs
+## Install dependencies
 
-Install yargs as a production dependency, as it will _not_ be included in the pack (not yet compatible):
 ```bash
-$ npm install --save yargs
+$ npm install --save-dev webpack webpack-cli yargs
 ```
 
-## Javascript setup (see below for typescript)
+Additional dependencies for typescript users:
 
-Install dev dependencies:
 ```bash
-$ npm install --save-dev webpack webpack-cli
+$ npm install --save-dev ts-loader typescript @types/yargs
 ```
+
+## Sample program
 
 Create `src/index.js`:
 ```js
@@ -21,39 +21,14 @@ const yargs = require('yargs')
 console.log(yargs.parse())
 ```
 
-Create `webpack.config.js`:
-```js
-const path = require('path');
-
-module.exports = {
-  entry: './src/index.js',
-  // Keep yargs as an external dependency
-  externals: {
-    'yargs': 'commonjs yargs'
-  },
-  output: {
-    path: path.resolve(__dirname, 'dist'),
-    filename: 'index.js'
-  },
-  target: 'node'
-}
-```
-
-## Typescript setup
-
-Install dev dependencies:
-```bash
-$ npm install --save-dev ts-loader typescript webpack webpack-cli @types/yargs
-```
-
-Create `src/index.ts`:
+Or for typescript users, `src/index.ts`:
 ```ts
 import * as yargs from 'yargs';
 
 console.log(yargs.parse());
 ```
 
-Create `tsconfig.json`:
+along with its `tsconfig.json`:
 ```json
 {
   "compilerOptions": {
@@ -62,16 +37,36 @@ Create `tsconfig.json`:
 }
 ```
 
+## Webpack configuration
+
 Create `webpack.config.js`:
 ```js
 const path = require('path');
 
 module.exports = {
-  entry: './src/index.ts',
-  // Keep yargs as an external dependency
-  externals: {
-    'yargs': 'commonjs yargs'
+  entry: './src/index.js',
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    filename: 'index.js'
   },
+  target: 'node'
+}
+```
+
+For typescript users, replace :
+
+```js
+module.exports = {
+  entry: './src/index.js',
+  ...
+}
+```
+
+by:
+
+```js
+module.exports = {
+  entry: './src/index.ts',
   module: {
     rules: [
       {
@@ -82,14 +77,10 @@ module.exports = {
       }
     ]
   },
-  output: {
-    path: path.resolve(__dirname, 'dist'),
-    filename: 'index.js'
-  },
   resolve: {
     extensions: ['.ts', '.js'],
   },
-  target: 'node'
+  ...
 }
 ```
 
@@ -99,9 +90,72 @@ module.exports = {
 $ ./node_modules/.bin/webpack --mode=production
 ```
 
+Webpack will emit the following warnings, which can be safely ignored as long as you do not use yargs functions performing dynamic module imports, such as `commandDir` or `config`:
+
+```
+WARNING in ./node_modules/yargs/index.js 12:39-46
+Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
+ @ ./src/index.js
+
+WARNING in ./node_modules/yargs/yargs.js 370:33-40
+Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
+ @ ./node_modules/yargs/index.js
+ @ ./src/index.js
+
+WARNING in ./node_modules/yargs/yargs.js 525:83-90
+Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
+ @ ./node_modules/yargs/index.js
+ @ ./src/index.js
+
+WARNING in ./node_modules/require-main-filename/index.js 2:25-32
+Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
+ @ ./node_modules/yargs/yargs.js
+ @ ./node_modules/yargs/index.js
+ @ ./src/index.js
+
+WARNING in ./node_modules/yargs/lib/apply-extends.js 42:24-55
+Critical dependency: the request of a dependency is an expression
+ @ ./node_modules/yargs/yargs.js
+ @ ./node_modules/yargs/index.js
+ @ ./src/index.js
+
+WARNING in ./node_modules/yargs/lib/apply-extends.js 57:82-105
+Critical dependency: the request of a dependency is an expression
+ @ ./node_modules/yargs/yargs.js
+ @ ./node_modules/yargs/index.js
+ @ ./src/index.js
+
+WARNING in ./node_modules/yargs/node_modules/yargs-parser/index.js 534:21-48
+Critical dependency: the request of a dependency is an expression
+ @ ./node_modules/yargs/yargs.js
+ @ ./node_modules/yargs/index.js
+ @ ./src/index.js
+```
+
 ## Run
 
 ```bash
+$ rm -rf node_modules
 $ node dist/index.js
 { _: [], '$0': 'dist/index.js' }
+```
+
+## Keeping yargs as an external dependency
+
+If you choose not to include yargs in the pack, to keep it as an external dependency (to get rid of webpack warnings, for example), you need then to install it as a production dependency, no longer a development one:
+
+```bash
+$ npm install --save yargs
+```
+
+And to add the following lines in `webpack.config.js`:
+
+```js
+module.exports = {
+  ...
+  externals: {
+    'yargs': 'commonjs yargs'
+  },
+  ...
+}
 ```

--- a/docs/webpack.md
+++ b/docs/webpack.md
@@ -1,0 +1,100 @@
+# Webpack usage examples
+
+## Setup
+
+### Javascript
+
+Create `src/index.js`:
+```ts
+const yargs = require('yargs')
+
+console.log(yargs.parse())
+```
+
+Create `webpack.config.js`:
+```js
+const path = require('path');
+const nodeExternals = require('webpack-node-externals');
+
+module.exports = {
+  entry: './src/index.js',
+  externals: [nodeExternals()],
+  mode: 'production',
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    filename: 'index.js'
+  },
+  target: 'node'
+}
+```
+
+Install dependencies (all as dev-dependencies, as prod will not need any dependency):
+```bash
+$ npm install --save-dev webpack webpack-cli webpack-node-externals yargs
+```
+
+### Typescript
+
+Create `src/index.ts`:
+```ts
+import * as yargs from 'yargs';
+
+console.log(yargs.parse());
+```
+
+Create `tsconfig.json`:
+```json
+{
+  "compilerOptions": {
+    "sourceMap": true
+  }
+}
+```
+
+Create `webpack.config.js`:
+```js
+const path = require('path');
+const nodeExternals = require('webpack-node-externals');
+
+module.exports = {
+  entry: './src/index.ts',
+  externals: [nodeExternals()],
+  mode: 'production',
+  module: {
+    rules: [
+      {
+        test: /\.ts$/,
+        use: [
+          'ts-loader',
+        ]
+      }
+    ]
+  },
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    filename: 'index.js'
+  },
+  resolve: {
+    extensions: ['.ts', '.js'],
+  },
+  target: 'node'
+}
+```
+
+Install dependencies (all as dev-dependencies, as prod will not need any dependency):
+```bash
+$ npm install --save-dev ts-loader typescript webpack webpack-cli webpack-node-externals @types/yargs yargs
+```
+
+## Build
+
+```bash
+$ ./node_modules/.bin/webpack
+```
+
+## Run
+
+```bash
+$ node dist/index.js
+{ _: [], '$0': 'dist\\index.js' }
+```

--- a/docs/webpack.md
+++ b/docs/webpack.md
@@ -49,6 +49,10 @@ module.exports = {
     path: path.resolve(__dirname, 'dist'),
     filename: 'index.js'
   },
+  stats: {
+    // Ignore warnings due to yarg's dynamic module loading
+    warningsFilter: [/node_modules\/yargs/]
+  },
   target: 'node'
 }
 ```
@@ -90,72 +94,10 @@ module.exports = {
 $ ./node_modules/.bin/webpack --mode=production
 ```
 
-Webpack will emit the following warnings, which can be safely ignored as long as you do not use yargs functions performing dynamic module imports, such as `commandDir` or `config`:
-
-```
-WARNING in ./node_modules/yargs/index.js 12:39-46
-Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
- @ ./src/index.js
-
-WARNING in ./node_modules/yargs/yargs.js 370:33-40
-Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
- @ ./node_modules/yargs/index.js
- @ ./src/index.js
-
-WARNING in ./node_modules/yargs/yargs.js 525:83-90
-Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
- @ ./node_modules/yargs/index.js
- @ ./src/index.js
-
-WARNING in ./node_modules/require-main-filename/index.js 2:25-32
-Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
- @ ./node_modules/yargs/yargs.js
- @ ./node_modules/yargs/index.js
- @ ./src/index.js
-
-WARNING in ./node_modules/yargs/lib/apply-extends.js 42:24-55
-Critical dependency: the request of a dependency is an expression
- @ ./node_modules/yargs/yargs.js
- @ ./node_modules/yargs/index.js
- @ ./src/index.js
-
-WARNING in ./node_modules/yargs/lib/apply-extends.js 57:82-105
-Critical dependency: the request of a dependency is an expression
- @ ./node_modules/yargs/yargs.js
- @ ./node_modules/yargs/index.js
- @ ./src/index.js
-
-WARNING in ./node_modules/yargs/node_modules/yargs-parser/index.js 534:21-48
-Critical dependency: the request of a dependency is an expression
- @ ./node_modules/yargs/yargs.js
- @ ./node_modules/yargs/index.js
- @ ./src/index.js
-```
-
 ## Run
 
 ```bash
 $ rm -rf node_modules
 $ node dist/index.js
 { _: [], '$0': 'dist/index.js' }
-```
-
-## Keeping yargs as an external dependency
-
-If you choose not to include yargs in the pack, to keep it as an external dependency (to get rid of webpack warnings, for example), you need then to install it as a production dependency, no longer a development one:
-
-```bash
-$ npm install --save yargs
-```
-
-And to add the following lines in `webpack.config.js`:
-
-```js
-module.exports = {
-  ...
-  externals: {
-    'yargs': 'commonjs yargs'
-  },
-  ...
-}
 ```

